### PR TITLE
174654525 fill blank question detail

### PR DESCRIPTION
**Pull request by eireland**
Adds ability to show question prompts for scaffolded questions and fill in the blank questions. 
Added mock data to reflect that prompts are part of the data sent to the report as part of the question map.
Refactored how questions are rendered in the question area so each type is now a separate component like the answers.

### CODE REVIEW

File: css/portal-dashboard/questions/multiple-choice-question.less [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+    margin-right: 8px;
+  }
+  .correct {
+    font-weight: bold;
+  }
+}
+
+.questionText {
+  margin: 0px 20px 10px 20px;
+}
```
Comment (mklewandowski) [inline css/portal-dashboard/questions/multiple-choice-question.less:27]: newline

File: css/portal-dashboard/question-area.less [hunk 2]
```diff
@@ -20,9 +20,10 @@
     display: flex;
     flex-direction: row;
     align-items: center;
-    flex: 1;
+    flex-grow: 2;
```
Comment (mklewandowski) [inline css/portal-dashboard/question-area.less:23]: Why is this needed?  `flex: 1` sets `flex-grow` to 1.  This change will make this box grow at 2 times the rate of other flex boxes in the same container.  I'm wondering if the better approach is to leave this as 1, and then make it so the `rightIcons` class doesn't grow, by using `flex: auto` or setting a width and using `flex-grow: 0`.  I think you might get better results with that.

File: js/components/portal-dashboard/questions/multiple-choice-question.tsx [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+export default class MultipleChoiceQuestion extends PureComponent<IProps> {
+  render() {
+    const { question } = this.props;
+    const prompt = question?.get("prompt");
+    const choices: Map<string,any> = question?.get("choices") || [];
+
+    return (
+      <div className={css.questionText}>
+      {prompt && renderHTML(prompt)}
+      { choices.size > 0 ? choices.toArray().map(this.renderMultipleChoiceChoices(choices.toArray().length)) : "" }
```
Comment (mklewandowski) [inline js/components/portal-dashboard/questions/multiple-choice-question.tsx:20]: Tab in the content inside the div for consistency and readability.

File: js/containers/portal-dashboard/answer.tsx [hunk 4]
```diff
@@ -22,22 +22,21 @@ class Answer extends React.PureComponent<AnswerProps> {
     const scored = question && question.get("scored");
     const questionType = QuestionTypes.find(qt => qt.type === qtype && qt.scored === scored);
     const answerType = getAnswerType(answer, question);
```
Comment (mklewandowski) [inline js/containers/portal-dashboard/answer.tsx:24]: Removed unused `answerType` (just whack the entire line).

File: js/components/portal-dashboard/questions/multiple-choice-question.tsx [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+  private renderMultipleChoiceChoices = (numChoices: number) => (choices: Map<string, any>, i: number) => {
+    let multipleChoiceContent, multipleChoiceContentClass;
+    if (choices.get("correct")) {
+      multipleChoiceContentClass = `${css.mcContent} ${css.correct}`;
+      multipleChoiceContent = choices.get("content") + " (correct)";
+    }
+    else {
+      multipleChoiceContentClass = `${css.mcContent}`;
+      multipleChoiceContent = choices.get("content");
+    }
```
Comment (mklewandowski) [inline js/components/portal-dashboard/questions/multiple-choice-question.tsx:34]: Maybe replace the `let`  and the `if else` with:
```
    const multipleChoiceContent = `${choices.get("content")} ${choices.get("correct") ? "(correct)" : ""}`;
    const multipleChoiceContentClass = `${css.mcContent} ${choices.get("correct") ? css.correct : ""}`;
```

File: js/components/portal-dashboard/questions/multiple-choice-question.tsx [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+
+    return (
+      <div className={css.questionText}>
+      {prompt && renderHTML(prompt)}
+      { choices.size > 0 ? choices.toArray().map(this.renderMultipleChoiceChoices(choices.toArray().length)) : "" }
+      </div>
+    );
+  }
+
+  private renderMultipleChoiceChoices = (numChoices: number) => (choices: Map<string, any>, i: number) => {
```
Comment (mklewandowski) [inline js/components/portal-dashboard/questions/multiple-choice-question.tsx:25]: I don't think you need this function currying.  Can just be:
`private renderMultipleChoiceChoices = (choices: Map<string, any>, i: number) => {`

File: js/components/portal-dashboard/questions/multiple-choice-question.tsx (hunk 3 above)
Comment (mklewandowski) [reply js/components/portal-dashboard/questions/multiple-choice-question.tsx:20]: Can be rewritten as: 
`{ choices.toArray?.().map(this.renderMultipleChoiceChoices) }`

File: js/components/portal-dashboard/questions/multiple-choice-question.tsx [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+      multipleChoiceContentClass = `${css.mcContent} ${css.correct}`;
+      multipleChoiceContent = choices.get("content") + " (correct)";
+    }
+    else {
+      multipleChoiceContentClass = `${css.mcContent}`;
+      multipleChoiceContent = choices.get("content");
+    }
+    return (
+      <div className={css.choiceWrapper} key={`choices ${i}`}>
+        <div className={`${css.choiceIcon}`} />
```
Comment (mklewandowski) [inline js/components/portal-dashboard/questions/multiple-choice-question.tsx:37]: `<div className={css.choiceIcon} />`

File: js/components/portal-dashboard/questions/multiple-choice-question.tsx [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+        <div className={`${multipleChoiceContentClass}`}>
```
Comment (mklewandowski) [inline js/components/portal-dashboard/questions/multiple-choice-question.tsx:38]: `<div className={multipleChoiceContentClass}>`

File: js/components/portal-dashboard/questions/open-response-question.tsx [hunk 9]
```diff
@@ -0,0 +1,22 @@
...
+    const { question } = this.props;
+    const prompt = question?.get("prompt");
+
+    return (
+      <div className={css.questionText}>
+        {prompt && renderHTML(prompt)}
+      </div>
+    );
+  }
+}
```
Comment (mklewandowski) [inline js/components/portal-dashboard/questions/open-response-question.tsx:22]: I'll just note that these question components would be good choices to turn into functional components.

File: js/containers/portal-dashboard/question.tsx [hunk 10]
```diff
@@ -0,0 +1,46 @@
+import React from "react";
+import { Map } from "immutable";
+import MultipleChoiceQuestion from "../../components/portal-dashboard/questions/multiple-choice-question";
+import OpenResponseQuestion from "../../components/portal-dashboard/questions/open-response-question";
+import  ImageQuestion from "../../components/portal-dashboard/questions/image-question";
```
Comment (mklewandowski) [inline js/containers/portal-dashboard/question.tsx:5]: Extra space sneaked in here after `import`

File: js/containers/portal-dashboard/question.tsx [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+    return (
+      <div className={css.questionText} data-cy="current-question">
+          {this.renderQuestion(qtype)}
+      </div>
+    );
+  }
+
+  renderQuestion = (type: string) => {
+    const { question } = this.props;
+    const QuesitonComponent: any = {
```
Comment (mklewandowski) [inline js/containers/portal-dashboard/question.tsx:28]: `QuestionComponent`

File: js/containers/portal-dashboard/question.tsx [hunk 12]
```diff
@@ -0,0 +1,46 @@
...
+      "multiple_choice": MultipleChoiceQuestion,
+      "open_response": OpenResponseQuestion,
+      "image_question": ImageQuestion,
+      "iframe_interactive": IframeQuestion,
+    };
+    const QComponent = (question && QuesitonComponent[type]);
```
Comment (mklewandowski) [inline js/containers/portal-dashboard/question.tsx:34]: `QuestionComponent`

File: js/containers/portal-dashboard/question.tsx [hunk 13]
```diff
@@ -0,0 +1,46 @@
...
+      return (
+        <div>Question type not supported.</div>
+      );
+    }
+    else {
+      return (
+        <QComponent question={question} />
+      );
+    }
+  }
```
Comment (mklewandowski) [inline js/containers/portal-dashboard/question.tsx:45]: I would remove the if/else and two returns and use:
```
    return (
      <React.Fragment>
      { QComponent 
        ? <QComponent question={question} />
        : <div>Question type not supported.</div>
      }
      </React.Fragment>
    );
```

**Review by mklewandowski — CHANGES_REQUESTED** [review]
Looks good!  Almost there.  I made comments for some code clean-up.  I would also investigate the `flex-grow: 2` and try to keep that as `flex: 1` while instead changing the `rightIcons` class to be a static width (use `flex-grow: 0` and set width).  It should prevent some unexpected results and perform better.  Let me know if that makes sense.  The other changes should be in my comments.

File: js/containers/portal-dashboard/question.tsx (hunk 13 above)
Comment (kswenson) [reply js/containers/portal-dashboard/question.tsx:45]: Since there's only a single child you shouldn't need the `<React.Fragment>` either.

Comment (mklewandowski) [reply js/containers/portal-dashboard/question.tsx:45]: Yes. Can be:
```
    return (
       QComponent
        ? <QComponent question={question} />
        : <div>Question type not supported.</div>
    );
```

File: css/portal-dashboard/question-area.less (hunk 2 above)
Comment (eireland) [reply css/portal-dashboard/question-area.less:23]: Set the width for the left side title text, and positioned the right side icons.

File: js/components/portal-dashboard/level-viewer.tsx [hunk 14]
```diff
@@ -8,7 +8,7 @@ import LaunchIcon from "../../../img/svg-icons/launch-icon.svg";
 import css from "../../../css/portal-dashboard/level-viewer.less";
 
 // from level-viewer.less
-const progressWidth = 120;
+const progressWidth = 116;
```
Comment (mklewandowski) [inline js/components/portal-dashboard/level-viewer.tsx:11]: Per my comments in Slack, I'm voting for this to be 117.

File: css/portal-dashboard/questions/multiple-choice-question.less (hunk 1 above)
Comment (mklewandowski) [reply css/portal-dashboard/questions/multiple-choice-question.less:27]: Might still want to add a newline at the end of this file to be consistent.  I know we haven't enforced this as much in CSS modules, but seems like a good practice.

File: js/containers/portal-dashboard/answer.tsx [hunk 15]
```diff
@@ -1,7 +1,7 @@
 import React from "react";
 import { getAnswersByQuestion } from "../../selectors/report-tree";
 import { connect } from "react-redux";
-import { getAnswerType, getAnswerIconId, AnswerProps } from "../../util/answer-utils";
+import { getAnswerType, AnswerProps } from "../../util/answer-utils";
```
Comment (mklewandowski) [inline js/containers/portal-dashboard/answer.tsx:4]: `getAnswerType` is no longer used and throws a lint warning.  We can remove it from the imports.

File: js/components/portal-dashboard/questions/multiple-choice-question.tsx [hunk 16]
```diff
@@ -0,0 +1,50 @@
...
+
+  //   return (
+  //     <div className={css.choiceWrapper} key={`choices ${i}`}>
+  //       <div className={css.choiceIcon} />
+  //       <div className={multipleChoiceContentClass}>
+  //         {multipleChoiceContent}
+  //       </div>
+  //     </div>
+  //   );
+  // }
```
Comment (mklewandowski) [inline js/components/portal-dashboard/questions/multiple-choice-question.tsx:49]: Remove obsolete code.

File: js/components/portal-dashboard/questions/multiple-choice-question.tsx [hunk 17]
```diff
@@ -0,0 +1,50 @@
...
+        </div>
+      );
+    };
+
+    return (
+      <div className={css.questionText}>
+        {prompt && renderHTML(prompt)}
+        { choices.toArray?.().map(renderMultipleChoiceChoices) }
+      </div>
+    );
```
Comment (mklewandowski) [inline js/components/portal-dashboard/questions/multiple-choice-question.tsx:34]: Might be clearer to try to collapse this so all the render is in one return.  I know that's something that @kswenson mentioned a couple times about many of our React class components - that we have sprawling render functions that are hard to follow.

File: js/components/portal-dashboard/question-area.tsx [hunk 18]
```diff
@@ -17,23 +17,20 @@ export class QuestionArea extends React.PureComponent<IProps>{
...
-    const showDivider = drawingPrompt && prompt;
 
     return (
       <div className={`${css.questionArea} ${hideQuestion ? css.hidden : ""}`}>
         <div className={css.questionTypeHeader}>
           <div className={css.leftTitle}>
             <QuestionIcon className={`${css.icon} ${css.questionTypeIcon}`} />
             <span className={css.questionTypeTitle} data-cy="question-title">
-              {typeText}
+              {type === "iframe_interactive"? (interactiveName? interactiveName : typeText) : typeText}
```
Comment (mklewandowski) [inline js/components/portal-dashboard/question-area.tsx:33]: I think this will be more clear if in line 23 you have:
`const interactiveName = type === "iframe_interactive" && currentQuestion?.get("name");`

And then here in the render you just use::
`{ interactiveName || typeText }`

This way, you're more fully defining `interactiveName` based on both the type and the question and the render portion becomes more clear and doesn't require two ternaries.

File: js/components/portal-dashboard/question-area.tsx [hunk 19]
```diff
@@ -17,23 +17,20 @@ export class QuestionArea extends React.PureComponent<IProps>{
     const { currentQuestion, hideQuestion, useMinHeight } = this.props;
     const teacherEditionButtonClasses = css.teacherEditionButton;
     const teacherEditionBadge = css.teacherEditionBadge;
-    const prompt = currentQuestion?.get("prompt");
     const type = currentQuestion?.get("type");
     const scored = currentQuestion?.get("scored");
     const typeText = type && type.replace(/_/gm, ' ');
+    const interactiveName = currentQuestion?.get("name");
```
Comment (mklewandowski) [inline js/components/portal-dashboard/question-area.tsx:23]: Per my comment below:
`const interactiveName = type === "iframe_interactive" && currentQuestion?.get("name");`

**Review by mklewandowski — CHANGES_REQUESTED** [review]
Almost there!  Just a few final tweaks:
- add newline at the end of .less module
- eliminate double ternary in render in `question-area.tsx` (just change two lines per my comments)
- remove obsolete code in `multiple-choice-question.tsx` 
- refactor render in `multiple-choice-question.tsx` per my suggestion
- remove unused import in `answer.tsx` that throws lint warning

File: js/components/portal-dashboard/questions/multiple-choice-question.tsx (hunk 17 above)
Comment (mklewandowski) [reply js/components/portal-dashboard/questions/multiple-choice-question.tsx:34]: Here, this should do the trick:
```
    return (
      <div className={css.questionText}>
        { prompt && renderHTML(prompt) }
        { choices.toArray?.().map((choices: Map<string, any>, i: number) => {
            const multipleChoiceContent = `${choices.get("content")} ${choices.get("correct") ? "(correct)" : ""}`;
            const multipleChoiceContentClass = `${css.mcContent} ${choices.get("correct") ? css.correct : ""}`;
            return (
              <div className={css.choiceWrapper} key={`choices ${i}`}>
                <div className={css.choiceIcon} />
                <div className={multipleChoiceContentClass}>
                  {multipleChoiceContent}
                </div>
              </div>
            );
          })
        }
      </div>
    );
```

Comment (mklewandowski) [reply js/components/portal-dashboard/questions/multiple-choice-question.tsx:34]: Replace your `return` in the functional component with the above.  Then you can remove the `renderMultipleChoiceChoices`.  It will be easier to read and might make code climate happy!

File: js/components/portal-dashboard/questions/iframe-question.tsx [hunk 20]
```diff
@@ -0,0 +1,22 @@
+import React, { PureComponent } from "react";
```
Comment (mklewandowski) [inline js/components/portal-dashboard/questions/iframe-question.tsx:1]: Remove unused `PureComponent` import.

File: js/components/portal-dashboard/questions/image-question.tsx [hunk 21]
```diff
@@ -0,0 +1,26 @@
+import React, { PureComponent } from "react";
```
Comment (mklewandowski) [inline js/components/portal-dashboard/questions/image-question.tsx:1]: Remove unused `PureComponent` import.

File: js/components/portal-dashboard/questions/open-response-question.tsx [hunk 22]
```diff
@@ -0,0 +1,20 @@
+import React, { PureComponent } from "react";
```
Comment (mklewandowski) [inline js/components/portal-dashboard/questions/open-response-question.tsx:1]: Remove unused `PureComponent` import.

**Review by mklewandowski — APPROVED** [review]
👍 Looks good!  Let's wait for it to build, and then bring it in!
